Sources/Sugar/Authentication/UserType.swift

```diff
@@ -36,6 +50,61 @@ extension UserType where
     }
 }
 
+extension UserType where
+    Self: Model,
+    Self: PasswordAuthenticatable,
+    Self.Database: QuerySupporting,
+    Self.Registration: HasReadableUser
+{
+    public static func preRegister(with registration: Registration, on worker: DatabaseConnectable) throws -> Future<Void> {
+        return try Self.query(on: worker)
+            .filter(Self.usernameKey == registration.username)
+            .first()
+            .nil(or: AuthenticationError.usernameAlreadyExists)
+    }
+}
+
+extension UserType where
+    Self: Model,
+    Self: PasswordAuthenticatable,
+    Self.Database: QuerySupporting,
+    Self.Update: HasUpdatableUsername,
+    Self.Update: HasUpdatablePassword,
+    Self: HasPassword
+{
+    public func preUpdate(with update: Update, on worker: DatabaseConnectable) throws -> Future<Void> {
+        if update.password != nil && (update.username == nil && update.oldPassword == nil) {
+            throw Abort(.internalServerError)
+        }
+
+        if let oldPassword = update.oldPassword {
+            guard try verify(oldPassword) else {
+                throw AuthenticationError.incorrectOldPassword
+            }
+        }
+
+        if let username = update.username {
+            guard let password = update.password, try verify(password) else {
+                throw AuthenticationError.incorrectPassword
+            }
+
+            return try Self.query(on: worker)
+                .filter(Self.usernameKey == username)
+                .first()
+                .try { user in
+                    guard
+                        user == nil || user?[keyPath: Self.idKey] == self[keyPath: Self.idKey]
+                    else {
+                        throw AuthenticationError.usernameAlreadyExists
+                    }
+                }
+                .value(())
```

Comment:
has this changed from `transform`?

Reply:
No, I've updated it to use `transform` here (and removed the `value` on instances) and then renamed the static variant of `value` to `transform`, so that you're able to do `Future.transform(to: ..)`.